packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -528,12 +499,108 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
 
             IController(controller).withdrawUniPositionToken(data.vaultId);
 
-            _closeUniLp(data);
+            (uint256 wPowerPerpAmountInLp, ) = ControllerHelperUtil.closeUniLp(
+                nonfungiblePositionManager,
+                ControllerHelperDataType.closeUniLpParams({
+                    tokenId: data.tokenId,
+                    liquidity: data.liquidity,
+                    liquidityPercentage: data.liquidityPercentage,
+                    amount0Min: data.amount0Min,
+                    amount1Min: data.amount1Min
+                }),
+                isWethToken0
+            );
 
-            _checkPartialLpClose(data.vaultId, data.tokenId, data.liquidityPercentage);
+            // close short position using amounts collected from closing LP, withdraw collateralToWithdraw + deposited collateralToFlashloan
+            _closeShortWithAmountsFromLp(
+                data.vaultId,
+                wPowerPerpAmountInLp,
+                data.wPowerPerpAmountToBurn,
+                data.collateralToWithdraw.add(data.collateralToFlashloan),
+                data.limitPriceEthPerPowerPerp
+            );
+            ControllerHelperUtil.checkPartialLpClose(nonfungiblePositionManager, controller, data.vaultId, data.tokenId, data.liquidityPercentage);
+        } 
+        else if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_REBALANCE_VAULT_NFT) {
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            RebalanceVaultNftParams[] memory data = abi.decode(_calldata, (RebalanceVaultNftParams[]));
+            console.log(data.length);
+
+            for(uint256 i; i < data.length; i++) {
+                if (data[i].rebalanceVaultNftType == RebalanceVaultNftType.IncreaseLiquidity) {
+                    IController(controller).deposit{value: _amount}(data[i].vaultId);
+                    IController(controller).withdrawUniPositionToken(data[i].vaultId);
```

Comment:
If this is just for increasing liquidity, not sure we need to withdrawUniPositionToken().  Controller owns the NFT, so can't it just add liquidity without removing or a flash loan?

Reply:
I refactored here, so code changed. But basically it is everywhere needed IMO, to withdraw the UNI position to increase/decrease LP liquidity, as the call to uniswap is coming from the ControllerHelper not Controller, so if Controller own the NFT, it will not work. wyt ?